### PR TITLE
[feature] Improved support for STM32H7

### PR DESCRIPTION
**Pull request by Ant-ON**
This fix improves MCU detection (specifically H7). Patch will add a check of the current MCU instructions mode and switch to thumb mode. In some cases, MCU may wrongly be in ARM instruction mode.
Optimizes MCU firmware via gdb. Adds automatic detection of cache presence and breakpoints style.

I checked this changes on the debug boards that I have (F0/F1/F3/G4). Also this branch was checked for H7 in #1059.

These changes solve the problems #1063 and #1059 (partially)

### CODE REVIEW

File: src/stlink-lib/reg.h [hunk 1]
```diff
@@ -2,19 +2,60 @@
 #define STLINK_REG_H_
 
 #define STLINK_REG_CM3_CPUID           0xE000ED00
+
 #define STLINK_REG_CM3_FP_CTRL         0xE0002000
-#define STLINK_REG_CM3_FP_COMP0        0xE0002008
+#define STLINK_REG_CM3_FP_COMPn(n)    (0xE0002008 + n*4)
+#define STLINK_REG_CM7_FP_LAR          0xE0000FB0
+#define STLINK_REG_CM7_FP_LAR_KEY           0xC5ACCE55
```
Comment (Nightwalker-87) [inline src/stlink-lib/reg.h:9]: Please correct the indentation in this file to match the current formatting.
Don't use tabspacing as they vary between architectures.
(I believe this is the reason for the misalignment.)

File: CMakeLists.txt [hunk 2]
```diff
@@ -104,6 +109,7 @@ set(STLINK_HEADERS
         src/stlink-lib/md5.h
         src/stlink-lib/sg.h
         src/stlink-lib/usb.h
+		src/stlink-lib/helper.h
```
Comment (Nightwalker-87) [inline CMakeLists.txt:112]: Please correct the indentation in this file to match the current formatting.
Don't use tabspacing as they vary between architectures.
(I believe this is the reason for the misalignment.)

File: src/stlink-lib/reg.h (hunk 1 above)
Comment (Ant-ON) [reply src/stlink-lib/reg.h:9]: An additional shift is deliberately used here. There are no tab chars.
STLINK_REG_CM7_FP_LAR is address of register.
STLINK_REG_CM7_FP_LAR_KEY is value for LAR register.

File: CMakeLists.txt (hunk 2 above)
Comment (Ant-ON) [reply CMakeLists.txt:112]: This is where I really overlooked. I'll fix it soon.

Comment (Ant-ON) [reply CMakeLists.txt]: Solved by additional commit

File: CMakeLists.txt [hunk 3]
```diff
@@ -114,6 +120,7 @@ set(STLINK_SOURCE
         src/stlink-lib/md5.c
         src/stlink-lib/sg.c
         src/stlink-lib/usb.c
+		src/stlink-lib/helper.c
```
Comment (slyshykO) [inline CMakeLists.txt:123]: wrong indentation.

Comment (Ant-ON) [reply CMakeLists.txt:123]: I fixed it by addition commit 1bbdd74. It appears in the Commits tab, but its changes are not applied to the Files changed tab. I do not know why it is so.

Comment (slyshykO) [reply CMakeLists.txt:123]: you dit it for the helper.**h**, please do it for the helper.**c** also.

Comment (Ant-ON) [reply CMakeLists.txt]: oops, sorry. I fixed it too.

**Review by Nightwalker-87 — APPROVED** [review]

**Comment by Nightwalker-87** [comment]
@Ant-ON: Now, as our Travis CI is running again,
I found that this PR introduces some build failures when compiling with clang.
Can you focus on that? Thanks in advance and sorry for the late feedback, we had no indication for that during review.

**Comment by Ant-ON** [comment]
@Nightwalker-87 I saw a link to the problem in a nearby discussion. I will try to fix it as soon as possible.

**Comment by Nightwalker-87** [comment]
Eh yes, as this one has been merged already, but you can continue on the used branch `Ant-ON:try_h7_debug` after pulling recent changes from our current `develop` branch to that personal feature branch. I'd propose to name the new PR "Extended support for STM32H7 (bugfix for clang)" and we shall be fine. I'll wait with any further contributions until you submitted a working patch. :-)

**Pull request merged**